Simplify cms forms

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplifies cms forms
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | Simplifying Improve -> Design -> Pages. Everything must look the same aside from help now appearing under inputs instead as blue box, also some fields now will appear as required because they always were.

:notebook: **BC Break**
Backwards compatibility break introduced due to extension of TranslationAwareType by CmsPageCategoryType. This means if any module extends this type they will get an exception upon upgrading to PS version containing changes in this PR.

CleanHtmlValidator now has one dependancy. This means if any module extends this class they will get an exception upon upgrading to PS version containing changes in this PR.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/21243)
<!-- Reviewable:end -->

JevgenijVisockij committed:
Started CMS Form simplification

JevgenijVisockij committed:
Added todo

JevgenijVisockij committed:
Added todo for components

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php:
```diff
@@ -73,8 +74,12 @@ public function __construct(array $allCmsCategories, $isShopFeatureEnabled)
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $invalidCharactersForCatalogLabel = $this->trans('Invalid characters:', 'Admin.Notifications.Info') .  '<>;=#{}';
```
```suggestion
        $invalidCharactersForCatalogLabel = $this->trans('Invalid characters:', 'Admin.Global') .  '<>;=#{}';
```

Arman-Hosseini replied:
Hi @Julievrz ,
There is an extra dot in: `Admin..Global`

Julievrz replied:
Oops, thank you @Arman-Hosseini !

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php:
```diff
@@ -73,8 +74,12 @@ public function __construct(array $allCmsCategories, $isShopFeatureEnabled)
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $invalidCharactersForCatalogLabel = $this->trans('Invalid characters:', 'Admin.Notifications.Info') .  '<>;=#{}';
+        $invalidCharactersForNameLabel = $this->trans('Invalid characters:', 'Admin.Notifications.Info') . '<>={}';
```
```suggestion
        $invalidCharactersForNameLabel = $this->trans('Invalid characters:', 'Admin.Global') . '<>={}';
```

JevgenijVisockij committed:
Simplified CMS Category type

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml

JevgenijVisockij committed:
Changed formatted Text Area to use TranslateType

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php

Changed domain to an updated one.

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php

Changed domain to an updated one.

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
CS Fixer fixes

atomiix commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php:
```diff
@@ -85,24 +90,27 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                         ]),
                         new Length([
                             'max' => 64,
```
Maybe extract those `'max'` values into constantes?

JevgenijVisockij replied:
Makes sense, thanks! Added them.

atomiix reviewed (dismissed):
Just one comment but approving anyway!

Progi1984 reviewed (commented):
Agree with the feedback of @atomiix

JevgenijVisockij committed:
Added const for max length values in CmsPageCategoryType.php

JevgenijVisockij committed:
Removed unneeded friendly url max length const

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php:
```diff
@@ -34,19 +34,23 @@
+    private const META_TITLE_MAX_LENGTH = 255;
+    private const META_KEYWORDS_MAX_LENGTH = 255;
+    private const META_DESCRIPTION_MAX_LENGTH = 512;
 
```
I have added constants for max length values. While META_TITLE_MAX_LENGTH  and META_KEYWORDS_MAX_LENGTH have the same values, I can't find a connection between them, so I added 2 different constants.

PierreRambaud replied:
I see no reason to make these consts as private :thinking:

PierreRambaud replied:
The class is not final, it can be extended, it can also be used in a template, I recommend putting these kinds of consts in public

JevgenijVisockij replied:
Good point made them public.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php:
```diff
@@ -112,75 +125,83 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('friendly_url', TranslatableType::class, [
+                'label' => $this->trans('Friendly URL', 'Admin.Global'),
+                'help' => $this->trans('Only letters and the minus (-) character are allowed.', 'Admin.Catalog.Help'),
```
@JevgenijVisockij Regarding your comment in another PR (https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547865863) should we change the help text here too for "Unless the 'Accented URL' option is enabled (in Shop parameters > Traffic & SEO), only letters, numbers, underscores (_), and hyphens (-) are allowed." ?

JevgenijVisockij replied:
I think that makes complete sense! Added it.

Julievrz reviewed (commented):
A wording question following your comment in a previous PR. Thanks!

matks commented:
I'm afraid we need rebase now 😭

JevgenijVisockij commented:
> I'm afraid we need rebase now sob

It's done

atomiix commented:
@JevgenijVisockij you have still [a question](https://github.com/PrestaShop/PrestaShop/pull/22476#discussion_r555658847) from @Julievrz

JevgenijVisockij committed:
Added more clear help

khouloudbelguith commented:
Hi @JevgenijVisockij,

I found some issues:

**First issue**
1. Go to BO > Design > Pages page
2. Click on "Add new page category"
3. On this page > Click on `Help`
4. See error => redirection to another page: it shouldn't, it should be opened in the same page 

![image](https://user-images.githubusercontent.com/16067358/106878594-6a76b400-66da-11eb-8c3a-7b48f96e3469.png)

**Second issue**
1. Go to BO > Design > Pages page
2. Create two page-category cat1 & cat2
3. Sort by Position ASC 
4. Try to change the positions 
![image](https://user-images.githubusercontent.com/16067358/106879184-19b38b00-66db-11eb-9609-0d31fa21aa64.png)
5. Go to FO > sitemap > See, the category pages are not well ordered 
![image](https://user-images.githubusercontent.com/16067358/106880281-7794a280-66dc-11eb-9199-3ce1a2c9fab2.png)

**Third issue**
1. Go to BO > Design > Pages page
2. Create a page-category cat1 in the home category
3. Create a page 1 in the home category
4. In the `categories` section, filter by any field (A category page cat1 should be displayed)
5. In the `Pages in category "Home" ` section, filter by any field 
![image](https://user-images.githubusercontent.com/16067358/106881374-c0009000-66dd-11eb-997b-89cb4c5f50a7.png)
6.  Click on view for the cat1
7. See error => the filter is applied in all lists => NOK
![image](https://user-images.githubusercontent.com/16067358/106881497-e9212080-66dd-11eb-8c69-98a2cd4fe9d0.png)

**Fourth issue**

0. Disable debug mode
1. Go to BO > Design > Pages page
2. Click on "Add new page"
3. Set required fields
4. In the `Meta description` field > set a long-expression
5. Save
6. See error (unclear alert) 
![image](https://user-images.githubusercontent.com/16067358/106881960-78c6cf00-66de-11eb-8a68-5d17892d46ff.png)

**Fifth issue**

1. Disable debug mode
2. In the BO > Shop Parameters > General page >`Allow iframes on HTML fields` is disabled
3. Go to BO > Design > Pages page
4. Click on "Add new page"
5. Set required fields
6. In the page content > Add a video
7. Save
8. See error (Unclear alert) 
![image](https://user-images.githubusercontent.com/16067358/106882326-ec68dc00-66de-11eb-8ec3-172636584014.png)


**Sixth issue**
1. Go to BO > Design > Pages page
2. Click on "Add new page"
3. In the page content > Click on the `Bullet list` or click on the `Numbered list`
4. See an error in the console 
![image](https://user-images.githubusercontent.com/16067358/106882626-436eb100-66df-11eb-9522-242b3578c133.png)


Thank you!

JevgenijVisockij committed:
Fixed issue with CMS category not working

JevgenijVisockij commented:
Hey @khouloudbelguith thanks for Q&A! Still checking other issues but regarding second issue, it seemed to be a problem with sitemap rather then CMS form. So I created a separate issue and PR for it.

https://github.com/PrestaShop/PrestaShop/issues/23168
https://github.com/PrestaShop/PrestaShop/pull/23169

JevgenijVisockij commented:
Regarding the third issue.

You expect that each CMS category should have it's own cached filters.

So if we have Home -> Cat1. I filter in "Home" by name="Cat1", go to Cat1 it should no active filters and go back to "Home" it should still be name="Cat1". Is that correct?

As far as I can tell that's not the case, because each page currently have one set of cached filters. If we take "Category" as an example the same thing happens as here. Go to categories -> Filter by name="Clothes -> go to "Clothes" -> filter name="Clothes" still there. Maybe this should be an improvement instead or am I misunderstanding something?

JevgenijVisockij committed:
Fixed issue with too long description not producing the error

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageType.php:
```diff
@@ -172,6 +171,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                            'maxMessage' => $this->trans(
+                                'This field cannot be longer than %limit% characters',
+                                'Admin.Notifications.Error',
+                                ['%limit%' => 512]
```
Please use constant

JevgenijVisockij replied:
Added constants

Progi1984 reviewed (commented):
A small feedback

JevgenijVisockij committed:
Modief CleanHtmlValidator to take into account iframes as well

JevgenijVisockij committed:
Added constants for max lengths

JevgenijVisockij commented:
@khouloudbelguith 6th issue does not seem to be related to CMS page and I am not sure how to solve it, so I created a separate ticket for it.
https://github.com/PrestaShop/PrestaShop/issues/23213

So
I have fixed 1st, 5th and 6th.
Comment regarding 3rd issue here https://github.com/PrestaShop/PrestaShop/pull/22476#issuecomment-774671159 
2nd and 4th problems now have their own issues.

JevgenijVisockij committed:
Added test for clean html validation with iframes

Progi1984 commented on src/Core/ConstraintValidator/CleanHtmlValidator.php:
```diff
@@ -57,7 +67,8 @@ public function validate($value, Constraint $constraint)
         $containsJavascriptEvents = preg_match('/(' . $this->getJavascriptEvents() . ')[\s]*=/ims', $value);
 
-        if ($containsScriptTags || $containsJavascriptEvents) {
+        $iframe = !$this->allowIframe && preg_match('/<[\s]*(i?frame|form|input|embed|object)/ims', $value);
```
Please move the pattern to private const.

JevgenijVisockij replied:
Done

Progi1984 reviewed (commented):
A last comment

JevgenijVisockij committed:
Moved IFrame pattern to const

PierreRambaud commented on src/Core/ConstraintValidator/CleanHtmlValidator.php:
```diff
@@ -37,6 +37,18 @@
  */
 final class CleanHtmlValidator extends ConstraintValidator
 {
+    private const IFRAME_PATTERN = '/<[\s]*(i?frame|form|input|embed|object)/ims';
```
It's not only iframe pattern :thinking:

JevgenijVisockij replied:
Well this is what PS currently checks for when setting allow Iframe is turned off. 

But yeah I can see why name can be confusing. But I am not sure about best name.
Perhaps on of those?
ADVANCED_HTML_PATTERN
THIRD_PARTY_HTML_PATTERN
EMBEDDABLE_HTML_PATTERN
RISKY_HTML_PATTERN
What do you think?

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageType.php:
```diff
@@ -50,6 +49,11 @@
  */
 class CmsPageType extends TranslatorAwareType
 {
+    private const TITLE_MAX_CHARS = 255;
```
I see no reason to put this const in private

PierreRambaud replied:
The class is not final, it can be extend, if you put it in private, no one can use it

JevgenijVisockij replied:
Good point. Made all consts like this in both page and page category public.

PierreRambaud commented on src/PrestaShopBundle/Resources/config/services/core/constraint_validator.yml:
```diff
@@ -58,3 +58,11 @@ services:
+  prestashop.core.constraint_validator.clean_html_validator:
+      class: 'PrestaShop\PrestaShop\Core\ConstraintValidator\CleanHtmlValidator'
+      arguments:
+        - "@=service('prestashop.adapter.legacy.configuration').getBoolean('PS_ALLOW_HTML_IFRAME')"
```
It's a BC break :)
As to be documented

JevgenijVisockij replied:
Updated PR's description, do I need to document it somewhere else?

PierreRambaud commented on tests/Unit/Core/ConstraintValidator/CleanHtmlValidatorTest.php:
```diff
@@ -56,8 +56,20 @@ public function testItFailsWhenJavascriptEventsAreGiven()
 
+    public function testItFailsWhenIframeIsGiven()
+    {
+        $htmlTag = '<iframe src="catvideo.html" /></iframe>';
```
You should add more tests, regarding the condition, you have to test with form, object, embed, etc. With or without spaces  ̀<    iframe   >`, What about `<script>`?

JevgenijVisockij replied:
Added more tests.

<script> tag is already tested by original test and it fails regardless if iframe can be used or not. Or you meant something else?

PierreRambaud commented on tests/Unit/Core/ConstraintValidator/CleanHtmlWithIframeValidatorTest.php:
```diff
@@ -0,0 +1,75 @@
+{
+    public function testItFailsWhenScriptTagsAreGiven()
+    {
+        $scriptTag = '<script></script>';
```
Just notice you create a new test class for it.
I recommend you to don't create class only for tests. Please put them into `CleanHtmlValidatorTest.php`, and add a dataProvider if needed :)

JevgenijVisockij replied:
The reason I created 2 test classes is so it works cleanly with class they extend: ConstraintValidatorTestCase. 
It's has function createValidator and then using whatever it returns for validation. Both of those tests require the same validator but different parameters.

So I think having 2 test classes is cleaner than hacking in the second validator in the same class.

PierreRambaud reviewed (dismissed):
Great pull request.
SOmething is annoying me with the PATTERN_IFRAME which is not related to iframe only :sweat_smile:

JevgenijVisockij committed:
Made length constants public

JevgenijVisockij committed:
Adding more tests for cleanHtmlValidator

JevgenijVisockij committed:
Made tests look at different scenarios

JevgenijVisockij committed:
Added test to fail where iframe with spaces given

JevgenijVisockij committed:
Changed IFRAME_PATTERN to EMBEDDABLE_HTML_PATTERN

JevgenijVisockij committed:
Changed the way components are inited in cms

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js

JevgenijVisockij commented:
@PierreRambaud  Thanks for feedback! Addressed it.

atomiix commented on src/Core/ConstraintValidator/CleanHtmlValidator.php:
```diff
@@ -37,6 +37,18 @@
+    /**
+     * @var bool
+     */
+    private $allowIframe;
+
+    public function __construct(bool $allowIframe)
+    {
+        $this->allowIframe = $allowIframe;
+    }
```
I guess I should be renamed `$allowEmbeddableHtml` 😅

JevgenijVisockij replied:
Good point. I kept it as allowIframe because that's how it was spelled in other places. But I guess you gotta start naming things correctly at some point.

Renamed the variable

JevgenijVisockij committed:
Renamed allowIframe to allowEmbedabbleHtml

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageCategoryType.php:
```diff
@@ -73,8 +79,12 @@ public function __construct(array $allCmsCategories, $isShopFeatureEnabled)
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $invalidCharactersForCatalogLabel = $this->trans('Invalid characters:', 'Admin.Global') . '<>;=#{}';
+        $invalidCharactersForNameLabel = $this->trans('Invalid characters:', 'Admin.Global') . '<>={}';
```
```suggestion
        $invalidCharactersForCatalogLabel = $this->trans('Invalid characters:', 'Admin.Global') . '<>;=#{}';
        $invalidCharactersForNameLabel = $this->trans('Invalid characters:', 'Admin.Global') . '<>={}';
```

Could you move the string `'<>={}'` as constant ?

JevgenijVisockij replied:
Added contants from TypedRegexValidator

JevgenijVisockij committed:
USing constants instead of plain strings

khouloudbelguith commented:
Hi @JevgenijVisockij,

It is ok :heavy_check_mark: 
https://drive.google.com/file/d/1MMTq6JA7zd3UCPSURsrrV30x6i44EJSK/view?usp=sharing

Thank you!

Progi1984 commented:
Thanks @JevgenijVisockij & @khouloudbelguith
